// @bsv/message-box-client/dist/cjs/src/RemittanceAdapter.js
"use strict";
/**
 * RemittanceAdapter - Adapts MessageBoxClient to the CommsLayer interface
 *
 * This adapter bridges MessageBoxClient with the ts-sdk RemittanceManager by implementing
 * the CommsLayer interface. It handles the protocol differences between the two systems,
 * particularly around message body format (MessageBoxClient returns parsed objects,
 * RemittanceManager expects JSON strings).
 *
 * @example
 * ```typescript
 * import { RemittanceAdapter } from '@bsv/message-box-client'
 * import { RemittanceManager } from '@bsv/sdk'
 * import { MessageBoxClient } from '@bsv/message-box-client'
 * import { WalletClient } from '@bsv/sdk'
 *
 * const wallet = new WalletClient()
 * const messageBox = new MessageBoxClient({ walletClient: wallet })
 * const commsLayer = new RemittanceAdapter(messageBox)
 *
 * const manager = new RemittanceManager(
 *   {
 *     messageBox: 'remittance_inbox',
 *     remittanceModules: [new Brc29RemittanceModule()]
 *   },
 *   wallet,
 *   commsLayer
 * )
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemittanceAdapter = void 0;
/**
 * Adapter that implements the CommsLayer interface for MessageBoxClient
 *
 * This class wraps MessageBoxClient to provide compatibility with the RemittanceManager
 * communications interface. It handles format conversions, particularly ensuring message
 * bodies are properly stringified for the RemittanceManager protocol.
 */
class RemittanceAdapter {
    /**
     * Creates a new RemittanceAdapter
     * @param messageBox - The MessageBoxClient instance to adapt
     */
    constructor(messageBox) {
        this.messageBox = messageBox;
    }
    /**
     * Sends a message over the store-and-forward channel
     * @param args - Message parameters (recipient, messageBox, body)
     * @param hostOverride - Optional host override
     * @returns The transport message ID
     */
    async sendMessage(args, hostOverride) {
        const result = await this.messageBox.sendMessage({
            recipient: args.recipient,
            messageBox: args.messageBox,
            body: args.body
        }, hostOverride);
        return result.messageId;
    }
    /**
     * Sends a message over the live channel.
     * MessageBoxClient handles transport fallback internally (WebSocket -> HTTP).
     * @param args - Message parameters (recipient, messageBox, body)
     * @param hostOverride - Optional host override
     * @returns The transport message ID
     */
    async sendLiveMessage(args, hostOverride) {
        const result = await this.messageBox.sendLiveMessage({
            recipient: args.recipient,
            messageBox: args.messageBox,
            body: args.body
        }, hostOverride);
        return result.messageId;
    }
    /**
     * Lists pending messages for a message box
     *
     * Note: MessageBoxClient returns message bodies as parsed objects, but RemittanceManager
     * expects them as JSON strings. This method handles the conversion.
     *
     * @param args - List parameters (messageBox, optional host)
     * @returns Array of peer messages with stringified bodies
     */
    async listMessages(args) {
        const defaultRecipient = await this.messageBox.getIdentityKey();
        const messages = await this.messageBox.listMessages({
            messageBox: args.messageBox,
            host: args.host
        });
        return messages.map(msg => this.toRemittancePeerMessage(msg, args.messageBox, defaultRecipient));
    }
    /**
     * Acknowledges messages (deletes them from the server inbox)
     * @param args - Array of message IDs to acknowledge
     */
    async acknowledgeMessage(args) {
        // MessageBoxClient's acknowledgeMessage expects the same format
        await this.messageBox.acknowledgeMessage({ messageIds: args.messageIds });
    }
    /**
     * Starts a live listener and normalizes inbound messages to the remittance PeerMessage shape.
     */
    async listenForLiveMessages(args) {
        const defaultRecipient = await this.messageBox.getIdentityKey();
        await this.messageBox.listenForLiveMessages({
            messageBox: args.messageBox,
            overrideHost: args.overrideHost,
            onMessage: (msg) => {
                args.onMessage(this.toRemittancePeerMessage(msg, args.messageBox, defaultRecipient));
            }
        });
    }
    toRemittancePeerMessage(msg, fallbackMessageBox, fallbackRecipient) {
        var _a, _b;
        return {
            messageId: msg.messageId,
            sender: msg.sender,
            recipient: ((_a = msg.recipient) !== null && _a !== void 0 ? _a : fallbackRecipient),
            messageBox: (_b = msg.messageBox) !== null && _b !== void 0 ? _b : fallbackMessageBox,
            body: this.toBodyString(msg.body)
        };
    }
    toBodyString(body) {
        if (typeof body === 'string')
            return body;
        try {
            return JSON.stringify(body);
        }
        catch {
            return String(body);
        }
    }
}
exports.RemittanceAdapter = RemittanceAdapter;
