// @bsv/message-box-client/dist/cjs/src/__tests/RemittanceAdapter.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-env jest */
const globals_1 = require("@jest/globals");
const RemittanceAdapter_js_1 = require("../RemittanceAdapter.js");
describe('RemittanceAdapter', () => {
    const myIdentityKey = '02b463b8ef7f03c47fba2679c7334d13e4939b8ca30dbb6bbd22e34ea3e9b1b0e4';
    const senderKey = '03f5d7a10f8ac22f0785a54d7d30fd009a77da27812f4e2f4ac9327dfcb5f65f86';
    it('delegates sendMessage and returns the transport messageId', async () => {
        const messageBox = {
            sendMessage: globals_1.jest.fn().mockResolvedValue({
                status: 'success',
                messageId: 'http-mid'
            })
        };
        const adapter = new RemittanceAdapter_js_1.RemittanceAdapter(messageBox);
        const result = await adapter.sendMessage({
            recipient: senderKey,
            messageBox: 'remittance_inbox',
            body: '{"v":1}'
        }, 'https://override-host');
        expect(result).toBe('http-mid');
        expect(messageBox.sendMessage).toHaveBeenCalledWith({
            recipient: senderKey,
            messageBox: 'remittance_inbox',
            body: '{"v":1}'
        }, 'https://override-host');
    });
    it('delegates sendLiveMessage (live path) and returns the transport messageId', async () => {
        const messageBox = {
            sendLiveMessage: globals_1.jest.fn().mockResolvedValue({
                status: 'success',
                messageId: 'ws-mid'
            })
        };
        const adapter = new RemittanceAdapter_js_1.RemittanceAdapter(messageBox);
        const result = await adapter.sendLiveMessage({
            recipient: senderKey,
            messageBox: 'remittance_inbox',
            body: '{"v":1}'
        }, 'https://override-host');
        expect(result).toBe('ws-mid');
        expect(messageBox.sendLiveMessage).toHaveBeenCalledWith({
            recipient: senderKey,
            messageBox: 'remittance_inbox',
            body: '{"v":1}'
        }, 'https://override-host');
    });
    it('normalizes listMessages shape for remittance and forwards host', async () => {
        const messageBox = {
            getIdentityKey: globals_1.jest.fn().mockResolvedValue(myIdentityKey),
            listMessages: globals_1.jest.fn().mockResolvedValue([
                {
                    messageId: 'm1',
                    sender: senderKey,
                    created_at: '2026-01-01T00:00:00Z',
                    updated_at: '2026-01-01T00:00:00Z',
                    body: { kind: 'invoice', amount: 1 }
                },
                {
                    messageId: 'm2',
                    sender: senderKey,
                    recipient: '020202020202020202020202020202020202020202020202020202020202020202',
                    messageBox: 'custom_box',
                    created_at: '2026-01-01T00:00:00Z',
                    updated_at: '2026-01-01T00:00:00Z',
                    body: '{"kind":"receipt"}'
                }
            ])
        };
        const adapter = new RemittanceAdapter_js_1.RemittanceAdapter(messageBox);
        const result = await adapter.listMessages({
            messageBox: 'remittance_inbox',
            host: 'https://remote-host'
        });
        expect(messageBox.listMessages).toHaveBeenCalledWith({
            messageBox: 'remittance_inbox',
            host: 'https://remote-host'
        });
        expect(result).toEqual([
            {
                messageId: 'm1',
                sender: senderKey,
                recipient: myIdentityKey,
                messageBox: 'remittance_inbox',
                body: '{"kind":"invoice","amount":1}'
            },
            {
                messageId: 'm2',
                sender: senderKey,
                recipient: '020202020202020202020202020202020202020202020202020202020202020202',
                messageBox: 'custom_box',
                body: '{"kind":"receipt"}'
            }
        ]);
    });
    it('forwards live listener setup and normalizes inbound message shape', async () => {
        const onPaymentMessage = globals_1.jest.fn();
        let forwardedListener;
        const messageBox = {
            getIdentityKey: globals_1.jest.fn().mockResolvedValue(myIdentityKey),
            listenForLiveMessages: globals_1.jest.fn().mockImplementation(async ({ onMessage }) => {
                forwardedListener = onMessage;
            })
        };
        const adapter = new RemittanceAdapter_js_1.RemittanceAdapter(messageBox);
        await adapter.listenForLiveMessages({
            messageBox: 'remittance_inbox',
            overrideHost: 'https://ws-host',
            onMessage: onPaymentMessage
        });
        expect(messageBox.listenForLiveMessages).toHaveBeenCalledWith({
            messageBox: 'remittance_inbox',
            overrideHost: 'https://ws-host',
            onMessage: expect.any(Function)
        });
        forwardedListener === null || forwardedListener === void 0 ? void 0 : forwardedListener({
            messageId: 'live-1',
            sender: senderKey,
            created_at: '2026-01-01T00:00:00Z',
            updated_at: '2026-01-01T00:00:00Z',
            body: { kind: 'settlement' }
        });
        expect(onPaymentMessage).toHaveBeenCalledWith({
            messageId: 'live-1',
            sender: senderKey,
            recipient: myIdentityKey,
            messageBox: 'remittance_inbox',
            body: '{"kind":"settlement"}'
        });
    });
});
