// @bsv/sdk/dist/cjs/src/script/templates/SignatureUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeSignatureScope = computeSignatureScope;
exports.resolveSourceDetails = resolveSourceDetails;
exports.formatPreimage = formatPreimage;
const TransactionSignature_js_1 = __importDefault(require("../../primitives/TransactionSignature.js"));
const utils_js_1 = require("../../primitives/utils.js");
/**
 * Computes the signature scope flags from the given signing parameters.
 */
function computeSignatureScope(signOutputs, anyoneCanPay) {
    let signatureScope = TransactionSignature_js_1.default.SIGHASH_FORKID;
    if (signOutputs === 'all') {
        signatureScope |= TransactionSignature_js_1.default.SIGHASH_ALL;
    }
    if (signOutputs === 'none') {
        signatureScope |= TransactionSignature_js_1.default.SIGHASH_NONE;
    }
    if (signOutputs === 'single') {
        signatureScope |= TransactionSignature_js_1.default.SIGHASH_SINGLE;
    }
    if (anyoneCanPay) {
        signatureScope |= TransactionSignature_js_1.default.SIGHASH_ANYONECANPAY;
    }
    return signatureScope;
}
/**
 * Resolves and validates the source transaction details needed for signing.
 * Returns the resolved sourceTXID, sourceSatoshis, lockingScript, and otherInputs.
 */
function resolveSourceDetails(tx, inputIndex, providedSourceSatoshis, providedLockingScript) {
    const input = tx.inputs[inputIndex];
    const otherInputs = tx.inputs.filter((_, index) => index !== inputIndex);
    const sourceTXID = input.sourceTXID ?? input.sourceTransaction?.id('hex');
    if (sourceTXID == null || sourceTXID === undefined) {
        throw new Error('The input sourceTXID or sourceTransaction is required for transaction signing.');
    }
    if (sourceTXID === '') {
        throw new Error('The input sourceTXID or sourceTransaction is required for transaction signing.');
    }
    const sourceSatoshis = providedSourceSatoshis ??
        input.sourceTransaction?.outputs[input.sourceOutputIndex].satoshis;
    if (sourceSatoshis == null || sourceSatoshis === undefined) {
        throw new Error('The sourceSatoshis or input sourceTransaction is required for transaction signing.');
    }
    const lockingScript = providedLockingScript ??
        input.sourceTransaction?.outputs[input.sourceOutputIndex]
            .lockingScript;
    if (lockingScript == null) {
        throw new Error('The lockingScript or input sourceTransaction is required for transaction signing.');
    }
    return { sourceTXID, sourceSatoshis, lockingScript, otherInputs };
}
/**
 * Formats the transaction preimage for signing.
 */
function formatPreimage(params) {
    const { tx, inputIndex, signatureScope, sourceTXID, sourceSatoshis, lockingScript, otherInputs, inputSequence } = params;
    const input = tx.inputs[inputIndex];
    return TransactionSignature_js_1.default.format({
        sourceTXID,
        sourceOutputIndex: (0, utils_js_1.verifyNotNull)(input.sourceOutputIndex, 'input.sourceOutputIndex must have value'),
        sourceSatoshis,
        transactionVersion: tx.version,
        otherInputs,
        inputIndex,
        outputs: tx.outputs,
        inputSequence: inputSequence ?? (0, utils_js_1.verifyNotNull)(input.sequence, 'input.sequence must have value'),
        subscript: lockingScript,
        lockTime: tx.lockTime,
        scope: signatureScope
    });
}
