// @bsv/sdk/dist/cjs/src/transaction/http/BinaryFetchClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BinaryFetchClient = exports.BinaryNodejsHttpClient = void 0;
exports.binaryHttpClient = binaryHttpClient;
const NodejsHttpRequestUtils_js_1 = require("./NodejsHttpRequestUtils.js");
/**
 * Adapter for Node Https module to be used as HttpClient
 */
class BinaryNodejsHttpClient {
    constructor(https) {
        this.https = https;
    }
    async request(url, requestOptions) {
        return await (0, NodejsHttpRequestUtils_js_1.executeNodejsRequest)(this.https, url, requestOptions, (data) => Buffer.from(data));
    }
}
exports.BinaryNodejsHttpClient = BinaryNodejsHttpClient;
/**
 * Adapter for Node Https module to be used as HttpClient
 */
class BinaryFetchClient {
    constructor(fetch) {
        this.fetch = fetch;
    }
    async request(url, options) {
        const fetchOptions = {
            method: options.method,
            headers: options.headers,
            body: options.data
        };
        const res = await this.fetch(url, fetchOptions);
        const data = await res.text();
        return {
            ok: res.ok,
            status: res.status,
            statusText: res.statusText,
            data: data
        };
    }
}
exports.BinaryFetchClient = BinaryFetchClient;
function binaryHttpClient() {
    const noHttpClient = {
        async request(..._) {
            throw new Error('No method available to perform HTTP request');
        }
    };
    if (typeof window !== 'undefined' && typeof window.fetch === 'function') {
        // Use fetch in a browser environment
        return new BinaryFetchClient(window.fetch.bind(window));
    }
    else if (typeof require !== 'undefined') {
        // Use Node https module
        // eslint-disable-next-line
        try {
            // eslint-disable-next-line @typescript-eslint/no-var-requires
            const https = require('https');
            return new BinaryNodejsHttpClient(https);
        }
        catch (e) {
            return noHttpClient;
        }
    }
    else {
        return noHttpClient;
    }
}
